rclpy/src/rclpy/action_server.cpp

```diff
@@ -0,0 +1,418 @@
+    &num_services);
+
+  if (RCL_RET_OK != ret) {
+    std::string error_text{"Failed to get number of entities for 'rcl_action_server_t'"};
+    throw rclpy::RCLError(error_text);
+  }
+
+  py::tuple result_tuple(5);
+  result_tuple[0] = py::int_(num_subscriptions);
+  result_tuple[1] = py::int_(num_guard_conditions);
+  result_tuple[2] = py::int_(num_timers);
+  result_tuple[3] = py::int_(num_clients);
+  result_tuple[4] = py::int_(num_services);
+  return result_tuple;
+}
+
+py::tuple
+ActionServer::is_ready(py::capsule pywait_set)
+{
+  auto wait_set = static_cast<rcl_wait_set_t *>(pywait_set);
+  bool is_goal_request_ready = false;
+  bool is_cancel_request_ready = false;
+  bool is_result_request_ready = false;
+  bool is_goal_expired = false;
+  rcl_ret_t ret = rcl_action_server_wait_set_get_entities_ready(
+    wait_set,
+    rcl_action_server_.get(),
+    &is_goal_request_ready,
+    &is_cancel_request_ready,
+    &is_result_request_ready,
+    &is_goal_expired);
+
+  if (RCL_RET_OK != ret) {
+    throw rclpy::RCLError("Failed to get number of ready entities for action server");
+  }
+
+  py::tuple result_tuple(4);
+  result_tuple[0] = py::bool_(is_goal_request_ready);
+  result_tuple[1] = py::bool_(is_cancel_request_ready);
+  result_tuple[2] = py::bool_(is_result_request_ready);
+  result_tuple[3] = py::bool_(is_goal_expired);
+  return result_tuple;
+}
+
+void
+ActionServer::add_to_waitset(py::capsule pywait_set)
+{
+  auto wait_set = static_cast<rcl_wait_set_t *>(pywait_set);
+  rcl_ret_t ret = rcl_action_wait_set_add_action_server(wait_set, rcl_action_server_.get(), NULL);
+  if (RCL_RET_OK != ret) {
+    std::string error_text{"Failed to add 'rcl_action_server_t' to wait set"};
+    throw rclpy::RCLError(error_text);
+  }
+}
+
+py::object
+ActionServer::process_cancel_request(
+  py::object pycancel_request, py::object pycancel_response_type)
+{
+  auto cancel_request = convert_from_py(pycancel_request);
+  rcl_action_cancel_request_t * cancel_request_type = static_cast<rcl_action_cancel_request_t *>(
+    cancel_request.get());
+  if (!cancel_request_type) {
+    throw py::error_already_set();
+  }
+
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+    rcl_action_server_.get(), cancel_request_type, &cancel_response);
+
+  if (RCL_RET_OK != ret) {
+    std::string error_text{"Failed to process cancel request: "};
+    error_text += rcl_get_error_string().str;
+    rcl_reset_error();
+
+    ret = rcl_action_cancel_response_fini(&cancel_response);
+    if (RCL_RET_OK != ret) {
+      error_text += ".  Also failed to cleanup response: ";
+      error_text += rcl_get_error_string().str;
+      rcl_reset_error();
+    }
+    throw std::runtime_error(error_text);
+  }
+
+  py::object return_value = convert_to_py(&cancel_response.msg, pycancel_response_type);
+
+  ret = rcl_action_cancel_response_fini(&cancel_response);
```

Comment:
@ahcorde consider putting this in an `scope_exit` block -- `convert_to_py` may throw.